exporter/exporterhelper/persistent_queue_test.go

```diff
@@ -0,0 +1,344 @@
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exporterhelper
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/extension/storage"
+)
+
+func createStorageExtension(_ string) storage.Extension {
+	// After having storage moved to core, we could leverage storagetest.NewTestExtension(nil, path)
+	return newMockStorageExtension()
+}
+
+func createTestQueue(extension storage.Extension, capacity int) *persistentQueue {
+	logger, _ := zap.NewDevelopment()
+
+	client, err := extension.GetClient(context.Background(), component.KindReceiver, config.ComponentID{}, "")
+	if err != nil {
+		panic(err)
+	}
+
+	wq := newPersistentQueue(context.Background(), "foo", capacity, logger, client, newTraceRequestUnmarshalerFunc(nopTracePusher()))
+	return wq
+}
+
+func createTestPersistentStorage(extension storage.Extension) persistentStorage {
+	logger, _ := zap.NewDevelopment()
+
+	client, err := extension.GetClient(context.Background(), component.KindReceiver, config.ComponentID{}, "")
+	if err != nil {
+		panic(err)
+	}
+
+	return newPersistentContiguousStorage(context.Background(), "foo", logger, client, newTraceRequestUnmarshalerFunc(nopTracePusher()))
+}
+
+func createTemporaryDirectory() string {
+	directory, err := ioutil.TempDir("", "persistent-test")
+	if err != nil {
+		panic(err)
+	}
+	return directory
+}
+
+func TestPersistentQueue_RepeatPutCloseReadClose(t *testing.T) {
+	path := createTemporaryDirectory()
+	defer os.RemoveAll(path)
+
+	traces := newTraces(5, 10)
+	req := newTracesRequest(context.Background(), traces, nopTracePusher())
+
+	for i := 0; i < 10; i++ {
+		ext := createStorageExtension(path)
+		ps := createTestPersistentStorage(ext)
+		require.Equal(t, 0, ps.size())
+
+		// Put two elements
+		err := ps.put(req)
+		require.NoError(t, err)
+		err = ps.put(req)
+		require.NoError(t, err)
+
+		err = ext.Shutdown(context.Background())
+		require.NoError(t, err)
+
+		// TODO: when replacing mock with real storage, this could actually be uncommented
+		// ext = createStorageExtension(path)
+		// ps = createTestPersistentStorage(ext)
+
+		// The first element should be already picked by loop
+		require.Eventually(t, func() bool {
+			return ps.size() == 1
+		}, 500*time.Millisecond, 10*time.Millisecond)
+
+		// Lets read both of the elements we put
+		readReq := <-ps.get()
```

Comment:
You might want to consider putting this in a `select` with `time.After(<timeout>)`+test_failure. Otherwise if something fails because the channel doesn't produce anything the test will get stuck and wait for global test timeout (default 10m)